libraries/mbed/targets/hal/TARGET_NXP/TARGET_LPC81X/us_ticker.c

```diff
@@ -17,77 +17,89 @@
 #include "us_ticker_api.h"
 #include "PeripheralNames.h"
 
-#define US_TICKER_TIMER_IRQn     SCT_IRQn
+//New, using MRT instead of SCT, needed to free up SCT for PWM
+//Ported from LPC824 libs
+static int us_ticker_inited = 0;
+static int ticker_expired = 0;
+int MRT_Clock_MHz;
 
-int us_ticker_inited = 0;
+#define US_TICKER_TIMER_IRQn     MRT_IRQn
 
 void us_ticker_init(void) {
-    if (us_ticker_inited) return;
-    us_ticker_inited = 1;
-    
-    // Enable the SCT clock
-    LPC_SYSCON->SYSAHBCLKCTRL |= (1 << 8);
-    
-    // Clear peripheral reset the SCT:
-    LPC_SYSCON->PRESETCTRL |= (1 << 8);
-    
-    // Unified counter (32 bits)
-    LPC_SCT->CONFIG |= 1;
-    
-    // halt and clear the counter
-    LPC_SCT->CTRL_L |= (1 << 2) | (1 << 3);
-    
-    // System Clock (12)MHz -> us_ticker (1)MHz
-    LPC_SCT->CTRL_L |= ((SystemCoreClock/1000000 - 1) << 5);
     
-    // unhalt the counter:
-    //    - clearing bit 2 of the CTRL register
-    LPC_SCT->CTRL_L &= ~(1 << 2);
+    if (us_ticker_inited)
+        return;
+
+    us_ticker_inited = 1;
+
+    // Calculate MRT clock value (MRT has no prescaler)
+    MRT_Clock_MHz = (SystemCoreClock / 1000000);
+
+    // Enable the MRT clock
+    LPC_SYSCON->SYSAHBCLKCTRL |= (1 << 10);
+
+    // Clear peripheral reset the MRT
+    LPC_SYSCON->PRESETCTRL |= (1 << 7);
+
+    // Force load interval value (Bit 0-30 is interval value, Bit 31 is Force Load bit)     
+    LPC_MRT->INTVAL0 = 0xFFFFFFFFUL;
+    // Enable Ch0 interrupt, Mode 0 is Repeat Interrupt
+    LPC_MRT->CTRL0 = (0x0 << 1) | (0x1 << 0);
+
+    // Force load interval value (Bit 0-30 is interval value, Bit 31 is Force Load bit)     
+    LPC_MRT->INTVAL1 = 0x80000000UL;
+    // Disable ch1 interrupt, Mode 0 is Repeat Interrupt
+    LPC_MRT->CTRL1 = (0x0 << 1) | (0x0 << 0);
     
+    // Set MRT interrupt vector
     NVIC_SetVector(US_TICKER_TIMER_IRQn, (uint32_t)us_ticker_irq_handler);
     NVIC_EnableIRQ(US_TICKER_TIMER_IRQn);
 }
 
+//TIMER0 is used for us ticker and timers (Timer, wait(), wait_us() etc)
 uint32_t us_ticker_read() {
+    
     if (!us_ticker_inited)
         us_ticker_init();
-    
-    return LPC_SCT->COUNT_U;
+
+    // Generate ticker value
+    // MRT source clock is SystemCoreClock (30MHz) and MRT is a 31-bit countdown timer
+    // Calculate expected value using number of expired times to mimic a 32bit timer @ 1 MHz
+    return (0x7FFFFFFFUL - LPC_MRT->TIMER0)/MRT_Clock_MHz + (ticker_expired * (0x80000000UL/MRT_Clock_MHz));
```

Comment:
I agree with scissors that the second division can be precalculated. In
fact, the current LPC824 code, which is identical, does that but is actually
wrong since it uses a hardcoded #define value of 30MHz instead of the
variable MRT_Clock_MHz. That works until someone decides to run the lpc824
at any other frequency but 30MHz.

 

Regarding computation: the MRT has only 31 bits instead of 32 as expected
for the us_ticker. In addition the clockrate is not 1 MHz but 30 MHz (ie
systemcoreclock). So there are 30 counts for 1 us. Hence the division. 

 

(ticker_expired * (0x80000000UL/MRT_Clock_MHz)

The variable ticker_expired keeps track of the number of 31bits overflows
(counted by TIMER0) and the multiplication/division corrects that back to us
counts.

 

(0x7FFFFFFFUL - LPC_MRT->TIMER0)/MRT_Clock_MHz

The counter is a 31bit downcounter from 7FFFFFFF so correct to actual
count-up value and correct for 30 counts per us. 

 

Added up these 2 parts result in current us time returned as 32 bits.

 

I agree its not really pretty but needed to support the PWMs which seems
valuable. The computation is taking place when a ticker expires or when you
are doing a wait() and reading the time in a loop. 

 

Regards,

Wim

  _____  

From: mbed-devel@googlegroups.com [mailto:mbed-devel@googlegroups.com] On
Behalf Of Martin Kojtal
Sent: vrijdag 13 februari 2015 9:35
To: mbedmicro/mbed
Cc: mbednotifications
Subject: Re: [mbed] Update us_ticker.c (#905)

 

In libraries/mbed/targets/hal/TARGET_NXP/TARGET_LPC81X/us_ticker.c
<https://github.com/mbedmicro/mbed/pull/905#discussion_r24650331> :

>      if (!us_ticker_inited)
>          us_ticker_init();
> -    
> -    return LPC_SCT->COUNT_U;
> +
> +    // Generate ticker value
> +    // MRT source clock is SystemCoreClock (30MHz) and MRT is a 31-bit
countdown timer
> +    // Calculate expected value using number of expired times to mimic a
32bit timer @ 1 MHz
> +    return (0x7FFFFFFFUL - LPC_MRT->TIMER0)/MRT_Clock_MHz +
(ticker_expired * (0x80000000UL/MRT_Clock_MHz));

Seems like that changing this from SCT to MRT has some tradeoffs. Wim, can
you share the assembly for calculating the return value here?

-
Reply to this email directly or view it on
<https://github.com/mbedmicro/mbed/pull/905/files#r24650331>  GitHub.
<https://github.com/notifications/beacon/AFAC1xTCWlo9s8AWLe6wIXjqOQZp2Qunks5
nra6XgaJpZM4DfFlc.gif> 


-- 
You received this message because you are subscribed to the Google Groups
"mbed-devel" group.
To unsubscribe from this group and stop receiving emails from it, send an
email to mbed-devel+unsubscribe@googlegroups.com.
For more options, visit https://groups.google.com/d/optout.